lib/python/treadmill_aws/garbage_collector.py

```diff
@@ -0,0 +1,75 @@
+"""AWS specific treadmill garbage collector
+"""
+
+import logging
+from datetime import datetime
+from datetime import timedelta
+
+import jmespath
+
+from treadmill import admin
+from treadmill import context
+from treadmill_aws.awscontext import GLOBAL
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def _valid_ec2_instances():
+    try:
+        return _valid_ec2_instances.cache
+    except:
+        pass
+
+    instance_name = jmespath.compile(
+        "Reservations[].Instances[].[Tags[?Key=='Name'].Value][][]")
+
+    _LOGGER.info('Fetching valid instances from AWS')
+
+    _valid_ec2_instances.cache = set(instance_name.search(
+        GLOBAL.ec2.describe_instances(
+            Filters=[{"Name": "instance-state-name",
+                      "Values": ["running", "pending", "rebooting",
+                                 "stopping", "stopped"]}])))
+
+    _LOGGER.debug("%d valid instances found", len(_valid_ec2_instances.cache))
+
+    return _valid_ec2_instances.cache
+
+
+def ldap():
+    """Remove terminated ec2 instances from LDAP"""
+    _LOGGER.info('LDAP cleanup started')
+    admin_srv = admin.Server(context.GLOBAL.ldap.conn)
+    # One hour grace time for servers to bootstrap and register
+    one_hour_ago = datetime.now().utcnow() - timedelta(seconds=3600)
+    ldap_servers = set(
+        server[1].get('server').pop()
+        for server in admin_srv.admin.search(
+            search_base='ou=servers,%s' % admin_srv.admin.root_ou,
+            search_filter="(&(server=*)(createTimestamp<=%s))" %
+            one_hour_ago.strftime("%Y%m%d%H%M%SZ"),
+            attributes=["server"]))
+
+    for hostname in ldap_servers - _valid_ec2_instances():
+        _LOGGER.info('removing %s from LDAP', hostname)
+        admin_srv.delete(hostname)
+    _LOGGER.info('LDAP cleanup complete')
+
+
+def ipa():
+    """Remove terminated ec2 instances from IPA.
```

Comment:
This function performs an implicit selection of hosts- while elegant, it is brittle. I'd prefer explicitly selecting State:terminated hosts from EC2, and to exclude hosts younger than an hour as in the ldap() function.

Reply:
thinking about it - terminated hosts will eventually disappear, right? so why bother? there is no rush removing them, this is gc. so it can be if in ec2 - valid, simple as that. no?

Reply:
removed filter:
instances = set(instance_name.search(GLOBAL.ec2.describe_instances()))